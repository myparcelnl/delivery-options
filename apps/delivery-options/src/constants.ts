import {ShipmentOptionName} from '@myparcel/constants';

export const FIELD_DELIVERY_MOMENT = 'deliveryMoment';

export const FIELD_DELIVERY_DATE = 'deliveryDate';

export const FIELD_SHIPMENT_OPTIONS = 'shipmentOptions';

export const SHOWN_SHIPMENT_OPTIONS = [ShipmentOptionName.Signature, ShipmentOptionName.OnlyRecipient] as const;
