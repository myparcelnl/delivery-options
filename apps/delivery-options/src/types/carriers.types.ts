import {type ComputedRef} from 'vue';
import {
  type CarrierIdentifier,
  type CarrierSettings,
  type CarrierSettingsKey,
  type ConfigKey,
  type DeliveryOptionsConfig,
  type SupportedDeliveryTypeName,
  type SupportedPackageTypeName,
  type SupportedShipmentOptionName,
} from '@myparcel-do/shared';
import {type Carrier} from '@myparcel/sdk';

export type ResolvedCarrier = Carrier & {
  identifier: CarrierIdentifier;
  pickupCountries: ComputedRef<Set<string>>;
  deliveryCountries: ComputedRef<Set<string>>;
  deliveryTypes: ComputedRef<Set<SupportedDeliveryTypeName>>;
  packageTypes: ComputedRef<Set<SupportedPackageTypeName>>;
  shipmentOptions: ComputedRef<Set<SupportedShipmentOptionName>>;
  features: ComputedRef<Set<string>>;
  hasDelivery: ComputedRef<boolean>;
  hasPickup: ComputedRef<boolean>;
  get<Key extends ConfigKey | CarrierSettingsKey>(
    key: Key,
    defaultValue?: NonNullable<Key extends ConfigKey ? DeliveryOptionsConfig[Key] : CarrierSettings[Key]>,
  ): Key extends ConfigKey ? DeliveryOptionsConfig[Key] : CarrierSettings[Key];
};
