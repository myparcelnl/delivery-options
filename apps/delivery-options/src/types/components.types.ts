import {type DeliveryOptionsOutput, type InputDeliveryOptionsConfiguration} from '@myparcel-do/shared';

export interface DeliveryOptionsProps {
  configuration?: InputDeliveryOptionsConfiguration;
}

export type DeliveryOptionsEmits = (event: 'update', values: DeliveryOptionsOutput) => void;
