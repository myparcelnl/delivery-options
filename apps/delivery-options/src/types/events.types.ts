import {type InputDeliveryOptionsConfiguration} from '@myparcel-do/shared';

export interface IncomingEventDetail extends InputDeliveryOptionsConfiguration {
  selector?: string;
}
