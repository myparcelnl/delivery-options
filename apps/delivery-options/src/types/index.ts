export * from './components.types';
export * from './form.types';
export * from './map.types';
