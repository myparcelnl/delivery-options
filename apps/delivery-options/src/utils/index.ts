export * from './calculateCutoffTime';
export * from './calculateDropOffDays';
export * from './createCarrierMarkerIcon';
export * from './createGetDeliveryOptionsParameters';
export * from './createNextDate';
export * from './createUtcDate';
export * from './getConfigFromWindow';
export * from './getConfigPriceKey';
export * from './getDefaultAddress';
export * from './getDeliveryTypePrice';
export * from './getResolvedCarrier';
export * from './getResolvedDeliveryType';
export * from './getResolvedValue';
export * from './hasSlotContent';
export * from './language';
export * from './padStart';
export * from './parseJson';
export * from './showDeveloperInfo';
export * from './stringToDate';
export * from './toTime';
