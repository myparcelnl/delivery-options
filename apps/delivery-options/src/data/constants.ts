import {ShipmentOptionName} from '@myparcel/constants';

export const FIELD_HOME_OR_PICKUP = 'homeOrPickup';

export const HOME_OR_PICKUP_HOME = 'home';

export const FIELD_DELIVERY_MOMENT = 'deliveryMoment';

export const FIELD_DELIVERY_DATE = 'deliveryDate';

export const FIELD_SHIPMENT_OPTIONS = 'shipmentOptions';

export const HOME_OR_PICKUP_PICKUP = 'pickup';

export const FIELD_PICKUP_LOCATION = 'pickupLocation';

export const SHOWN_SHIPMENT_OPTIONS = [ShipmentOptionName.Signature, ShipmentOptionName.OnlyRecipient] as const;

export const MAP_MARKER_CLASS_ACTIVE = 'active';

export const MAP_MARKER_CLASS_PREFIX = 'mp__carrier-marker';

export const SHOWN_OPENING_HOURS = 3;

export const OUTPUT_EVENT_DEBOUNCE_DELAY = 50;
