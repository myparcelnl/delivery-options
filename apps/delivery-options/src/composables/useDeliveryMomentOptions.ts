import {type ComputedRef, computed} from 'vue';
import {
  type SelectOption,
  PACKAGE_TYPE_DEFAULT,
  DELIVERY_TYPE_DEFAULT,
  type SupportedShipmentOptionName,
} from '@myparcel-do/shared';
import {getDeliveryTypePrice} from '../utils';
import {useConfigStore} from '../stores';
import {SHOWN_SHIPMENT_OPTIONS} from '../data';
import {useResolvedDeliveryMoments} from './useResolvedDeliveryMoments';
import {useLanguage} from './useLanguage';
import {useActiveCarriers} from './useActiveCarriers';

export const useDeliveryMomentOptions = (): ComputedRef<SelectOption[]> => {
  const config = useConfigStore();
  const deliveryMoments = useResolvedDeliveryMoments();
  const activeCarriers = useActiveCarriers();

  const {translate} = useLanguage();

  return computed(() => {
    if (PACKAGE_TYPE_DEFAULT !== config.packageType) {
      return activeCarriers.value
        .filter((carrier) => carrier.hasDelivery.value && carrier.allowedPackageTypes.value.has(config.packageType))
        .map((carrier) => {
          return {
            carrier: carrier.name,
            label: translate(`package_type_${config.packageType}`),
            price: getDeliveryTypePrice(DELIVERY_TYPE_DEFAULT, carrier.name),
            value: JSON.stringify({
              carrier: carrier.name,
              date: null,
              deliveryType: DELIVERY_TYPE_DEFAULT,
              packageType: config.packageType,
              shipmentOptions: [],
              time: null,
            }),
          };
        });
    }

    return deliveryMoments.value.map((option) => {
      return {
        carrier: option.carrier,
        label: option.time,
        price: getDeliveryTypePrice(option.deliveryType, option.carrier),
        value: JSON.stringify({
          time: option.time,
          carrier: option.carrier,
          date: option.date,
          deliveryType: option.deliveryType,
          packageType: option.packageType,
          shipmentOptions: option.shipmentOptions.filter((option) =>
            SHOWN_SHIPMENT_OPTIONS.includes(option.name as SupportedShipmentOptionName),
          ),
        }),
      };
    });
  });
};
