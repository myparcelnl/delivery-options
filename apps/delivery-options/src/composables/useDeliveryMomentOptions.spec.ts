import {type ComputedRef} from 'vue';
import {describe, it, expect, beforeEach} from 'vitest';
import {setActivePinia, createPinia} from 'pinia';
import {flushPromises} from '@vue/test-utils';
import {
  type SupportedPackageTypeName,
  type SelectOption,
  KEY_CONFIG,
  CarrierSetting,
  ConfigSetting,
  KEY_CARRIER_SETTINGS,
} from '@myparcel-do/shared';
import {CarrierName, PackageTypeName} from '@myparcel/constants';
import {parseJson} from '../utils';
import {type SelectedDeliveryMoment} from '../types';
import {mockSelectedDeliveryOptions, mockDeliveryOptionsConfig, waitForDeliveryOptions} from '../__tests__';
import {useDeliveryMomentOptions} from './useDeliveryMomentOptions';

const setup = async (packageType?: SupportedPackageTypeName): Promise<ComputedRef<SelectOption<string>[]>> => {
  mockDeliveryOptionsConfig({
    [KEY_CONFIG]: {
      [ConfigSetting.ShowDeliveryDate]: true,
      [CarrierSetting.AllowDeliveryOptions]: true,
      [CarrierSetting.AllowStandardDelivery]: true,
      [CarrierSetting.AllowEveningDelivery]: true,
      [CarrierSetting.AllowMorningDelivery]: true,
      [CarrierSetting.AllowSignature]: true,
      [CarrierSetting.AllowOnlyRecipient]: true,
      [CarrierSetting.PriceStandardDelivery]: 3,
      [KEY_CARRIER_SETTINGS]: {
        [CarrierName.PostNl]: {
          [CarrierSetting.PricePackageTypePackageSmall]: 6,
        },
        [`${CarrierName.PostNl}:123`]: {
          [CarrierSetting.PriceStandardDelivery]: 456,
          [CarrierSetting.PricePackageTypeMailbox]: 5,
          [CarrierSetting.PricePackageTypeDigitalStamp]: 4,
        },
        [CarrierName.DhlForYou]: {},
      },
      // TODO: allow optional key to be passed with undefined as value
      ...(packageType ? {[CarrierSetting.PackageType]: packageType} : {}),
    },
  });
  mockSelectedDeliveryOptions();

  const options = useDeliveryMomentOptions();
  await waitForDeliveryOptions();
  await flushPromises();

  return options;
};

describe('useDeliveryMomentOptions', () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it.each([
    PackageTypeName.Package,
    PackageTypeName.Mailbox,
    PackageTypeName.DigitalStamp,
    PackageTypeName.PackageSmall,
  ])('returns delivery moment options with package type %s', async (packageType) => {
    expect.assertions(2);

    const options = await setup(packageType);

    expect(options.value).toHaveLength(2);

    const resolved = options.value.map((option) => ({
      ...option,
      value: parseJson<SelectedDeliveryMoment>(option.value),
    }));

    expect(resolved).toMatchSnapshot();
  });
});
