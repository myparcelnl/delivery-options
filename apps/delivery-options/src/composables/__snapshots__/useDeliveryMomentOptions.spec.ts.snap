// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDeliveryMomentOptions > returns delivery moment options with package type digital_stamp 1`] = `
[
  {
    "carrier": "postnl",
    "label": {
      "key": "packageTypeDigitalStamp",
    },
    "price": 0,
    "value": {
      "carrier": "postnl",
      "date": null,
      "deliveryType": "standard",
      "packageType": "digital_stamp",
      "shipmentOptions": [],
      "time": null,
    },
  },
  {
    "carrier": "postnl:123",
    "label": {
      "key": "packageTypeDigitalStamp",
    },
    "price": 4,
    "value": {
      "carrier": "postnl:123",
      "date": null,
      "deliveryType": "standard",
      "packageType": "digital_stamp",
      "shipmentOptions": [],
      "time": null,
    },
  },
]
`;

exports[`useDeliveryMomentOptions > returns delivery moment options with package type mailbox 1`] = `
[
  {
    "carrier": "postnl",
    "label": {
      "key": "packageTypeMailbox",
    },
    "price": 0,
    "value": {
      "carrier": "postnl",
      "date": null,
      "deliveryType": "standard",
      "packageType": "mailbox",
      "shipmentOptions": [],
      "time": null,
    },
  },
  {
    "carrier": "postnl:123",
    "label": {
      "key": "packageTypeMailbox",
    },
    "price": 5,
    "value": {
      "carrier": "postnl:123",
      "date": null,
      "deliveryType": "standard",
      "packageType": "mailbox",
      "shipmentOptions": [],
      "time": null,
    },
  },
]
`;

exports[`useDeliveryMomentOptions > returns delivery moment options with package type package 1`] = `
[
  {
    "carrier": "postnl",
    "label": {
      "key": "deliveryStandardTitle",
    },
    "price": 3,
    "value": {
      "carrier": "postnl",
      "deliveryType": "standard",
      "packageType": "package",
      "shipmentOptions": [],
      "time": {
        "key": "deliveryStandardTitle",
      },
    },
  },
  {
    "carrier": "postnl:123",
    "label": {
      "key": "deliveryStandardTitle",
    },
    "price": 456,
    "value": {
      "carrier": "postnl:123",
      "deliveryType": "standard",
      "packageType": "package",
      "shipmentOptions": [],
      "time": {
        "key": "deliveryStandardTitle",
      },
    },
  },
]
`;

exports[`useDeliveryMomentOptions > returns delivery moment options with package type package_small 1`] = `
[
  {
    "carrier": "postnl",
    "label": {
      "key": "packageTypePackageSmall",
    },
    "price": 6,
    "value": {
      "carrier": "postnl",
      "date": null,
      "deliveryType": "standard",
      "packageType": "package_small",
      "shipmentOptions": [],
      "time": null,
    },
  },
  {
    "carrier": "postnl:123",
    "label": {
      "key": "packageTypePackageSmall",
    },
    "price": 0,
    "value": {
      "carrier": "postnl:123",
      "date": null,
      "deliveryType": "standard",
      "packageType": "package_small",
      "shipmentOptions": [],
      "time": null,
    },
  },
]
`;
